### PR TITLE
Self-contained whiteboards (almost)

### PR DESCRIPTION
**Pull request by mrMakaronka**

### CODE REVIEW

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java [hunk 1]
```diff
@@ -1,52 +1,15 @@
...
-        }
-        Field field = (Field) o;
-        return name.equals(field.name) && status == field.status;
-    }
-
-    @Override
-    public int hashCode() {
-        return Objects.hash(name, status);
-    }
+public record Field(String name, DataScheme schema) {
```
Comment (imakunin) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java:6]: let's use uniform codestyle for records
```
public record Field(
    String name,
    DataScheme schema
) {
    ...
}
```

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java [hunk 2]
```diff
@@ -1,52 +1,15 @@
...
 
     @Override
     public String toString() {
```
Comment (imakunin) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java:9]: you may wish to use autogenerated `toString` here

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/storage/WhiteboardStorage.java [hunk 3]
```diff
@@ -13,15 +12,9 @@
 
 public interface WhiteboardStorage {
 
-    void insertWhiteboard(String userId, Whiteboard whiteboard,
+    void insertWhiteboard(String userId, Whiteboard whiteboard, Instant ts,
```
Comment (imakunin) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/storage/WhiteboardStorage.java:15]: why `insert`?
I'd expect to have `createWhiteboard` method here

File: lzy/whiteboard/src/test/java/ai/lzy/whiteboard/ApiTest.java [hunk 4]
```diff
@@ -182,98 +172,337 @@ public void createAndGetWhiteboard() {
...
+                .build()));
+        assertTrue(sre.getMessage().startsWith("INVALID_ARGUMENT: Whiteboard must contain not blank ID"));
+
+        sre = Assert.assertThrows(StatusRuntimeException.class,
+            () -> whiteboardClient.registerWhiteboard(LWBS.RegisterWhiteboardRequest.newBuilder().setWhiteboard(
+                    LWB.Whiteboard.newBuilder()
+                        .setId(UUID.randomUUID().toString())
+                        .build())
+                .build()));
+        assertTrue(sre.getMessage().startsWith("INVALID_ARGUMENT: Whiteboard must contain not blank whiteboard name"));
```
Comment (imakunin) [inline lzy/whiteboard/src/test/java/ai/lzy/whiteboard/ApiTest.java:189]: it's a very bad practice to test error messages
you should test error code, not error message (the error code - is part of the API contract, error message is not)

even if you are testing error message, you should test error code too :)

File: pylzy/lzy/api/v1/snapshot.py [hunk 5]
```diff
@@ -27,84 +27,89 @@ class SnapshotEntry:
...
+            size = await self.__storage_client.size_in_bytes(entry.storage_uri)
             with tqdm(total=size, desc=f"Downloading {entry.name}", file=sys.stdout, unit='B', unit_scale=True,
                       unit_divisor=1024, colour=get_color()) as bar:
-                await self.storage_client.read(entry.storage_uri, cast(BinaryIO, f), progress=lambda x: bar.update(x))
+                await self.__storage_client.read(entry.storage_uri, cast(BinaryIO, f),
+                                                 progress=lambda x: bar.update(x))
                 f.seek(0)
-                res = self.__serializer_registry.find_serializer_by_type(entry.typ).deserialize(cast(BinaryIO, f))
+                res = self.__serializer_registry.find_serializer_by_type(entry.typ).deserialize(
+                    cast(BinaryIO, f))
```
Comment (imakunin) [inline pylzy/lzy/api/v1/snapshot.py:105]: ni: extra newline?

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/storage/WhiteboardStorage.java (hunk 3 above)
Comment (mrMakaronka) [reply lzy/whiteboard/src/main/java/ai/lzy/whiteboard/storage/WhiteboardStorage.java:15]: Actually, whiteboard is not created here, it is just added. Maybe `add` is better word

File: lzy/whiteboard/src/main/resources/db/whiteboard/migrations/V1_0__Create_tables.sql [hunk 6]
```diff
@@ -4,10 +4,11 @@ CREATE TABLE IF NOT EXISTS whiteboards (
     user_id              varchar(255)  NOT NULL,
     storage_name         varchar(255)  NOT NULL,
     storage_description  text          NOT NULL,
+    storage_uri          text          NOT NULL,
     namespace            varchar(255)  NOT NULL,
     whiteboard_status    varchar(255)  NOT NULL,
     created_at           timestamp     NOT NULL,
-    finalized_at         timestamp,
+    inserted_at           timestamp,
```
[the remark this answers is not in the document]
Comment (mrMakaronka) [reply lzy/whiteboard/src/main/resources/db/whiteboard/migrations/V1_0__Create_tables.sql]: Changed to `registered_at`

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java (hunk 1 above)
Comment (mrMakaronka) [reply lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java]: done

File: lzy/whiteboard/src/test/java/ai/lzy/whiteboard/ApiTest.java (hunk 4 above)
Comment (mrMakaronka) [reply lzy/whiteboard/src/test/java/ai/lzy/whiteboard/ApiTest.java]: agree, fixed

File: pylzy/lzy/api/v1/snapshot.py (hunk 5 above)
Comment (mrMakaronka) [reply pylzy/lzy/api/v1/snapshot.py]: fixed

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/grpc/WhiteboardService.java [hunk 7]
```diff
@@ -414,97 +300,101 @@ public void finalizeWhiteboard(FinalizeWhiteboardRequest request,
...
-
-        return true;
-    }
-
-    private static boolean validateLinkRequest(LinkFieldRequest request, StreamObserver<LinkFieldResponse> response) {
-        var methodName = "Link whiteboard field";
-
-        if (request.getWhiteboardId().isBlank()) {
-            return replyValidateError(methodName, "Request must contain not blank whiteboard id", response);
+        if (whiteboard.getStorage().getName().isBlank()) {
```
Comment (lindlind) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/grpc/WhiteboardService.java:383]: typo, getName twice

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/grpc/WhiteboardService.java [hunk 8]
```diff
@@ -414,97 +300,101 @@ public void finalizeWhiteboard(FinalizeWhiteboardRequest request,
...
+                        whiteboardId,
+                        requestWhiteboard.getName().isBlank() ? whiteboard.name() : requestWhiteboard.getName(),
+                        requestWhiteboard.getFieldsList().isEmpty() ? whiteboard.fields() :
+                            ProtoConverter.fromProto(requestWhiteboard.getFieldsList()),
+                        requestWhiteboard.getTagsList().isEmpty() ? whiteboard.tags() :
+                            new HashSet<>(requestWhiteboard.getTagsList()),
+                        newStorage,
+                        requestWhiteboard.getNamespace().isBlank() ? whiteboard.namespace() :
+                            requestWhiteboard.getNamespace(),
+                        requestWhiteboard.getStatus() == LWB.Whiteboard.Status.UNRECOGNIZED ? whiteboard.status() :
```
Comment (lindlind) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/grpc/WhiteboardService.java:324]: UNRECOGNIZED or UNSPECIFIED? UNRECOGNIZED = -1, UNSPECIFIED = 0.

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java [hunk 9]
```diff
@@ -1,52 +1,11 @@
...
-    }
-
-    public enum Status {
-        CREATED,
-        FINALIZED,
-    }
 
+public record Field(
+    String name,
+    DataScheme schema
```
Comment (lindlind) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Field.java:8]: There is scheme everywhere. Maybe here should be scheme too, not schema?

File: whiteboard-api/src/main/proto/ai/lzy/v1/whiteboard/whiteboard-service.proto [hunk 10]
```diff
@@ -9,11 +9,11 @@ option java_outer_classname = "LWBS";
 option java_package = "ai.lzy.v1.whiteboard";
 
 service LzyWhiteboardService {
+  rpc RegisterWhiteboard (RegisterWhiteboardRequest) returns (RegisterWhiteboardResponse);
+  rpc UpdateWhiteboard (UpdateWhiteboardRequest) returns (UpdateWhiteboardResponse);
+
   rpc Get (GetRequest) returns (GetResponse);
   rpc List (ListRequest) returns (ListResponse);
-  rpc CreateWhiteboard (CreateWhiteboardRequest) returns (CreateWhiteboardResponse);
-  rpc LinkField (LinkFieldRequest) returns (LinkFieldResponse);
```
Comment (lindlind) [inline whiteboard-api/src/main/proto/ai/lzy/v1/whiteboard/whiteboard-service.proto:15]: Where is link now?

File: whiteboard-api/src/main/proto/ai/lzy/v1/whiteboard/whiteboard-service.proto [hunk 11]
```diff
@@ -38,37 +38,20 @@ message ListResponse {
...
-message LinkFieldResponse {
-}
-
-// =========== Finalize ============
+// =========== Update ============
 
-message FinalizeWhiteboardRequest {
-  string whiteboardId = 1;
+message UpdateWhiteboardRequest {
+  Whiteboard whiteboard = 1;
```
Comment (lindlind) [inline whiteboard-api/src/main/proto/ai/lzy/v1/whiteboard/whiteboard-service.proto:53]: Are all fields mutable? Ok, id is not, it's key. But storage or createdAt?

File: whiteboard-api/src/main/proto/ai/lzy/v1/whiteboard/whiteboard.proto [hunk 12]
```diff
@@ -25,36 +25,15 @@ message Whiteboard {
   }
```
Comment (lindlind) [inline whiteboard-api/src/main/proto/ai/lzy/v1/whiteboard/whiteboard.proto:25]: There is no filalized request and there is update request. So what does finalized status mean?

If it's still necessary for user but not for WhiteboardService, why is it enum, not just string?

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Whiteboard.java [hunk 13]
```diff
@@ -46,13 +32,12 @@ public boolean equals(Object o) {
             return false;
         }
         Whiteboard that = (Whiteboard) o;
-        return id.equals(that.id) && name.equals(that.name) && fields.equals(that.fields) && tags.equals(that.tags) &&
-               storage.equals(that.storage) && namespace.equals(that.namespace) && status == that.status;
+        return id.equals(that.id);
```
Comment (lindlind) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Whiteboard.java:35]: With this change assertEquals will work incorrectly. We can create equalsAllFields method and use assertTrue(wb1.equalsAllFields(wb2)) instead.

Comment (lindlind) [reply lzy/whiteboard/src/main/java/ai/lzy/whiteboard/model/Whiteboard.java:35]: Do we really want equals to check only id?

File: lzy/whiteboard/src/main/java/ai/lzy/whiteboard/storage/WhiteboardStorage.java [hunk 14]
```diff
@@ -13,13 +12,7 @@
...
 public interface WhiteboardStorage {
 
-    void insertWhiteboard(String userId, Whiteboard whiteboard,
-                          @Nullable TransactionHandle transaction) throws SQLException;
-
-    void updateField(String whiteboardId, LinkedField field, @Nullable Instant finalizedAt,
-                     @Nullable TransactionHandle transaction) throws SQLException;
-
-    void finalizeWhiteboard(String whiteboardId, Instant finalizedAt,
+    void registerWhiteboard(String userId, Whiteboard whiteboard, Instant ts,
```
Comment (lindlind) [inline lzy/whiteboard/src/main/java/ai/lzy/whiteboard/storage/WhiteboardStorage.java:15]: ts == registeredAt? We can just name it as registeredAt.

File: lzy/whiteboard/src/main/resources/db/whiteboard/migrations/V1_0__Create_tables.sql [hunk 15]
```diff
@@ -4,10 +4,11 @@ CREATE TABLE IF NOT EXISTS whiteboards (
     user_id              varchar(255)  NOT NULL,
     storage_name         varchar(255)  NOT NULL,
     storage_description  text          NOT NULL,
+    storage_uri          text          NOT NULL,
     namespace            varchar(255)  NOT NULL,
     whiteboard_status    varchar(255)  NOT NULL,
     created_at           timestamp     NOT NULL,
-    finalized_at         timestamp,
+    registered_at        timestamp,
```
Comment (lindlind) [inline lzy/whiteboard/src/main/resources/db/whiteboard/migrations/V1_0__Create_tables.sql:11]: How can it be null?

**Review by imakunin — APPROVED** [review]

**Pull request merged**